tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h

```diff
@@ -155,8 +183,12 @@ class TrtShapeOptimizationProfile {
 
  private:
   // Set of input shape vetors that we collect during profile_generation_mode
-  std::unordered_set<std::vector<TensorShape>, VectorTensorShapeHasher>
-      input_shapes_;
+  // Let us consider one of the vectors shape_vec = input_shapes_[0]. The length
+  // of shape_vec is 2 x the number of engine inputs. The first half stores the
+  // shapes of the input tensors. The second half stores shape values if the
+  // tensor is compatible with a TRT shape tensor, otherwise it stores an empty
+  // shape.
+  std::vector<std::vector<TensorShape>> input_shapes_;
```

Comment:
I can see it is very convenient to append the shape_values to the end of the execution tensor shapes. I have two main questions regarding this:
(1) When selecting a cuda engine: Does OptimizationProfileConfig::IncludesShapes really need to be applied to the shape_values that are actually used as input tensors for reshape?  We also apply the check of OptimizationProfileConfig::IncludesShapes  to the shape values that are NOT used as input tensors for reshape, is that a problem? I tend to think it is a problem, and we will need to fix TrtShapeOptimizationProfile::GetProfileNumber and add test case for this.
(2) Is it possible to delay the copy from device to host of the shape_value in routine AppendShapeValues to TrtShapeOptimizationProfile::ConfigureBuilder? This can avoid unnecessary memcpy.

Reply:
Update: I have modified shape value handling the following way:
- `TrtShapeOptimizationProfile::CollectShapeValues` will store the shape values into the `actual_shape_values_` array. During profile collection, we maintain two vectors: `input_shapes_` and `input_shape_values_`. 
- After profile collection, in `InitProfiles` we concatenate the the vectors from `input_shapes_` and `input_shape_values_`. For both type of inputs we need to perform the same operations when we define the actual profiles, that is the motivation to treat them as a single `std::vector<nvinfer1::Dims>`.

1.b An `OptimizationProfileConfig` contains `2*n_input` values in its `min`, `max`, `opt` members, where `min[n_input + k]` ( 0 <= k < n_input) can describe a shape value input if. If input `k` is not a shape tensor, then the corresponding values will be initialized as `min = opt = max = nvinfer1::Dims{0, {}}`, This is because `CollectShapeValues` zero inits components of `actual_shape_values_` that do not correspond a input shape tensor. Such values will always pass the `IncludesShape` check called by GetProfileNumber.

1.c Once the network is initialized (or engine is loaded from file) we use `SetShapeTensorMask` to query which tensors are actually shape tensors. After this step `CollectShapeValues` will only copy data for inputs that are marked as input shape binding of the engine.

2. Synchronization to ensure that shape values are copied to host: currently it is done in `CollectShapeValues`. This makes the control flow clear. Alternatively we could postpone the sync call to `AppendShapeValues` and `SetInputShapeBinding`.